packages/account/src/BiconomySmartAccount.ts

```diff
@@ -74,38 +75,89 @@ export class BiconomySmartAccount extends SmartAccount implements IBiconomySmart
     try {
       this.isProviderDefined()
       this.isSignerDefined()
-      await this.setAccountIndex(accountIndex)
+      this.owner = await this.signer.getAddress()
       this.chainId = await this.provider.getNetwork().then((net) => net.chainId)
+      await this.setAccountIndex(accountIndex)
+      this.isInited = true
     } catch (error) {
-      console.error(`Failed to call init: ${error}`);
+      Logger.error(`Failed to call init: ${error}`);
       throw error
     }
 
     return this
   }
 
+  private isInitialized(): boolean{
+    if (!this.isInited)
+    throw new Error('BiconomySmartAccount is not initialized. Please call BiconomySmartAccount.init() before interacting with any other function')
+    return true
+  }
+
+  private setProxyContractState(){
+    const proxyInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: BICONOMY_IMPLEMENTATION_ADDRESSES[this.address],
+      contractAddress: this.address,
+      provider: this.provider
+    }
+    this.proxy = getProxyContractInstance(proxyInstanceDto)
+  }
+
+  private setEntryPointContractState(){
+    const _entryPointAddress = this.smartAccountInfo.entryPointAddress
+    this.setEntrypointAddress(_entryPointAddress)
+    const entryPointInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: ENTRYPOINT_ADDRESSES[_entryPointAddress],
+      contractAddress: _entryPointAddress,
+      provider: this.provider
+    }
+    this.entryPoint = getEntryPointContractInstance(entryPointInstanceDto)
+  }
+
+  private setFactoryContractState(){
+    const _factoryAddress = this.smartAccountInfo.factoryAddress
+    const factoryInstanceDto = {
+      smartAccountType: SmartAccountType.BICONOMY,
+      version: BICONOMY_FACTORY_ADDRESSES[_factoryAddress],
+      contractAddress: _factoryAddress,
+      provider: this.provider
+    }
+    this.factory = getFactoryContractInstance(factoryInstanceDto)
+  }
+
+  private async setContractsState() {
+    this.setProxyContractState()
+    this.setEntryPointContractState()
+    this.setFactoryContractState()
+  }
+
   async setAccountIndex(accountIndex: number): Promise<void> {
     this.accountIndex = accountIndex
     this.address = await this.getSmartAccountAddress(accountIndex)
-    this.proxy = await SmartAccount_v100__factory.connect(this.address, this.provider)
+    await this.setContractsState()
```

Comment:
Call these methods (setContractState() and setInitCode()) separately in init method... this method name is setAccountIndex and it should only set the account index.

Reply:
Then rename the method to `initializeAccountAtIndex` to accommodate the actions it's taking.

Reply:
resolved